### PR TITLE
Disable Request Waiter Timeout if not within fail fast threshold and Optimize SingleTimeout runnable to not capture callback to deal with the future.cancel behaviour in ScheduledExecutor

### PR DESCRIPTION
**Pull request by nizarm**
Disable Request Waiter Timeout if not within failfast threshold

### CODE REVIEW

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 1]
```diff
@@ -440,6 +444,11 @@ public boolean cancel()
     };
   }
 
+  private boolean isWaiterTimeoutWithinFastFailThreshold()
```
Comment (FreCap) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:447]: isWaiterTimeoutEnabled?

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 2]
```diff
@@ -440,6 +444,11 @@ public boolean cancel()
...
+  {
+    return _waiterTimeout >= MIN_WAITER_TIMEOUT && _waiterTimeout <= MAX_WAITER_TIMEOUT;
+  }
+
   @Override
   public void put(T obj)
   {
```
Comment (FreCap) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:454]: You can also fix the leak in the SingleTimeout

```
  public SingleTimeout(ScheduledExecutorService executor, long timeout, TimeUnit timeoutUnit, T item, Consumer<T> timeoutAction) 
  { 
    ArgumentUtil.ensureNotNull(item,"item"); 
    ArgumentUtil.ensureNotNull(timeoutAction,"timeoutAction"); 
 
    _item = new AtomicReference<>(item); 
    _future = executor.schedule(() -> { 
      T item1 = _item.getAndSet(null); 
      if (item1 != null) 
      { 
        try 
        { 
          timeoutAction.apply(item1); 
        } catch (Throwable e) 
        { 
          LOG.error("Failed to execute timeout action", e); 
        } 
      } 
    }, timeout, timeoutUnit); 
  } 
```

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 1 above)
Comment (FreCap) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:447]: isWaiterTimeoutEnabled?

Comment (FreCap) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:447]: within fast fail threshold can be a impl comment

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 2 above)
Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:455]: Done

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 3]
```diff
@@ -374,8 +374,12 @@ public Cancellable get(final Callback<T> callback)
           {
             if (_waiters.size() < _maxWaiters)
             {
+              if (isWaiterTimeoutWithinFastFailThreshold())
```
Comment (ssheng) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: This check should also exist within the constructor and offspring factory where it is passed in. We do address this in a separate PR if you'd like.

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 2 above)
Comment (ssheng) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:455]: SingleTime not part of this PR.

Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:455]: You mean - you dont want to include singletimeout? It is already present in this PR as a separate commit following this. I will squash these during merge time

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 3 above)
Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: Why this needs to be there in other places ? checking here should take care of all the cases right?

Comment (ssheng) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: We should not silently ignore configs passed in. If we could fail early in HttpClientFactory.Builder or the Offspring factory user can detect much earlier.

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 4]
```diff
@@ -840,15 +850,15 @@ public long getTime()
...
+      _timeout = new SingleTimeout<>(_timeoutExecutor, _waiterTimeout, TimeUnit.MILLISECONDS, callback, (callback1) -> {
 
         synchronized (_lock)
         {
           _waiters.remove(this);
           _statsTracker.incrementWaiterTimedOut();
         }
         LOG.debug("{}: failing waiter due to waiter timeout", _poolName);
-        callback.onError(
+        callback1.onError(
```
Comment (FreCap) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java:861]: I would rather not put a number in the callback variable, something like: callbackIfTimeout

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 3 above)
Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: Understood. I will create a separate PR for this.
